Implement AiProviderRegistry with comprehensive test suite

Ref34t opened the pull request:
Add core Provider Registry functionality that enables provider discovery and model selection based on requirements. This is a critical MVP component that unlocks the main AiClient functionality.

**Key Features:**
- Provider registration and validation
- Provider discovery by ID or class name
- Model metadata discovery with requirements matching
- Provider instance caching for performance
- Comprehensive error handling and validation

**Implementation Details:**
- Full API as specified in architecture documentation
- Follows WordPress coding standards and documentation practices
- Uses existing Provider DTOs (ProviderMetadata, ModelRequirements, etc.)
- Includes method existence validation for duck typing
- Comprehensive test coverage (13 tests, 28 assertions)

**Files Added:**
- src/Providers/AiProviderRegistry.php - Main registry implementation
- tests/unit/Providers/AiProviderRegistryTest.php - Full test suite
- tests/unit/Providers/MockProvider.php - Test provider implementation

**TODOs for Future Enhancement:**
- Integration with ProviderInterface when PR #35 merges
- Model metadata directory implementation
- Provider availability checking
- Model instantiation functionality

This implementation provides the foundation for provider-agnostic AI access and enables the next phase of MVP development.

Ref34t commented:
@felixarntz, kindly review my PR when you have an opportunity

Ref34t commented:
@felixarntz thank you 🙏 I really appreciate your detailed feedback. I will tackle your requests and will update the PR

Ref34t commented:
@felixarntz  Thank you for the detailed feedback! I have addressed all your requests:                                                                                                                                                      
                                                                                                                                                                                                                          
**Cleaned up unrelated changes:**                                                                                                                                                                                           
- Reverted ModelInterface to trunk version (method signatures)                                                                                                                                                                
- Reverted ModelMetadata to trunk version                                                                                                                                                                                    
- Reverted FileTest, ModelConfigTest, GenerativeAiResultTest to trunk versions                                                                                                                                                
- Removed GenerativeAiResultTest.php.bak backup file                                                                                                                                                                          
- Updated MockModel to match trunk method signatures     
                                                                                                                                                                                                                                                                                                                                                                                            
**All tests passing:** 543/543 tests pass                                                                                                                                                                                   
**Code quality:** PHPCS compliant with zero errors      
                                                                                                                                                                                                                                                                                                                                                                                                   
The PR now contains only the AiProviderRegistry implementation and related testing infrastructure. Ready for review when convenient.    
                                                                                                                                                                                                                              
 Thank you again for taking the time to review! 🙏

Ref34t commented:
**Question for next contribution:**  
 
@felixarntz  I've reviewed the current MVP progress in Issue #20 and noticed that you're actively working on the provider infrastructure (PR #39) and @JasonTheAdams is handling the HTTP infrastructure (PR #47).

For my next contribution, I'd like to focus on the remaining **client-facing API components** that aren't currently being worked on:

- **AiClient class** - Main public API entry point
- **PromptBuilder** - Fluent prompt building interface
- **MessageBuilder** - Message construction API
- **Utility classes** - RequirementsUtil and supporting logic

Which of these would be most valuable to tackle first? I'm particularly interested in the **AiClient class** since it would integrate directly with the Registry we just implemented and provide the main developer interface.

Any guidance on priority or architectural preferences would be greatly appreciated! 🙏

JasonTheAdams commented:
@Ref34t I'll be working on the `PromptBuilder` and `MesesageBuilder` today/tomorrow if you want to tackle the `AiClient` and utilities classes!

Thanks for jumping in with us! 🎉

JasonTheAdams commented on src/Providers/AiProviderRegistry.php:
```diff
@@ -0,0 +1,229 @@
+class AiProviderRegistry
+{
+    /**
+     * @var array<string, string> Mapping of provider IDs to class names.
```
```suggestion
     * @var array<string, class-string<ProviderInterface>> Mapping of provider IDs to class names.
```

Ref34t replied:
resolved [03310ba](https://github.com/WordPress/php-ai-client/pull/38/commits/03310ba68d62d46be002739601f81efd494627c6)

JasonTheAdams commented on src/Providers/AiProviderRegistry.php:
```diff
@@ -0,0 +1,229 @@
+     *
+     * @since n.e.x.t
+     *
+     * @param string $className The fully qualified provider class name.
```
```suggestion
     * @param class-string<ProviderInterface> $className The fully qualified provider class name implementing the ProviderInterface
```

Ref34t replied:
resolved https://github.com/WordPress/php-ai-client/pull/38/commits/8de52d28be6412c209720877e9d6a169613d1ea2

JasonTheAdams commented on src/Providers/AiProviderRegistry.php:
```diff
@@ -0,0 +1,229 @@
+        }
+
+        // Get provider metadata to extract ID (using static method from interface)
+        /** @var class-string<ProviderInterface> $className */
```
Shouldn't need this line if it's used in the parameter.

Ref34t replied:
resolved

JasonTheAdams commented on src/Providers/AiProviderRegistry.php:
```diff
@@ -0,0 +1,229 @@
+     *
+     * @since n.e.x.t
+     *
+     * @param string $idOrClassName The provider ID or class name to check.
```
```suggestion
     * @param string|class-string<ProviderInterface> $idOrClassName The provider ID or class name to check.
```

Ref34t replied:
resolved https://github.com/WordPress/php-ai-client/pull/38/commits/57dfe25e491ab836d4e201f5e29eea1129cfdda7

JasonTheAdams commented on src/Providers/AiProviderRegistry.php:
```diff
@@ -0,0 +1,229 @@
+     *
+     * @since n.e.x.t
+     *
+     * @param string $idOrClassName The provider ID or class name.
```
```suggestion
     * @param string|class-string<ProviderInterface> $idOrClassName The provider ID or class name.
```

Ref34t replied:
resolved https://github.com/WordPress/php-ai-client/pull/38/commits/8f96b26ebfcdf8ba1e87b419cc5347eef1d876ed

JasonTheAdams commented on src/Providers/AiProviderRegistry.php:
```diff
@@ -0,0 +1,229 @@
+    /**
+     * Gets the class name for a registered provider (handles both ID and class name input).
+     *
+     * @param string $idOrClassName The provider ID or class name.
+     * @return string The provider class name.
```
```suggestion
     * @param string|class-string<ProviderInterface> $idOrClassName The provider ID or class name.
     * @return class-string<ProviderInterface> The provider class name.
```

Ref34t replied:
resolved https://github.com/WordPress/php-ai-client/pull/38/commits/8a64e8a193f3d61483bdf92760785be496c16a29

JasonTheAdams commented on src/Providers/AiProviderRegistry.php:
```diff
@@ -0,0 +1,229 @@
+        $className = $this->resolveProviderClassName($idOrClassName);
+
+        // Use static method from ProviderInterface
+        /** @var class-string<ProviderInterface> $className */
```
In light of adjusting the `resolveProviderClassName()` return type we shouldn't need this.

Ref34t replied:
resolved https://github.com/WordPress/php-ai-client/pull/38/commits/489a4edf7752023491ec96fa415978bf33d98c16

JasonTheAdams commented on src/Providers/AiProviderRegistry.php:
```diff
@@ -0,0 +1,229 @@
+     *
+     * @since n.e.x.t
+     *
+     * @param string $idOrClassName The provider ID or class name.
```
```suggestion
     * @param string|class-string<ProviderInterface> $idOrClassName The provider ID or class name.
```

Ref34t replied:
https://github.com/WordPress/php-ai-client/pull/38/commits/8a64e8a193f3d61483bdf92760785be496c16a29

JasonTheAdams reviewed (changes requested):
Thanks for this great work, @Ref34t! I left some suggestions, mostly around types.

Also, can you please move the test mocks you made to the `tests/mocks` directory? All classes inside `tests/unit` should correspond directly with `src/`.

JasonTheAdams commented on src/Providers/AiProviderRegistry.php:
```diff
@@ -0,0 +1,229 @@
+
+        return $className;
+    }
+}
```
It's possible that PHPStan may not be happy here. The reason being that we're never actually validating that `$className` implements `ProviderInterface`. Technically we can assume that this is true since it would have to to pass `hasProvider`, but it may be worth adding a check in here if it does.

Ref34t replied:
refactored [212bab6](https://github.com/WordPress/php-ai-client/pull/38/commits/212bab69ad3c64aa78db5728e011d852442c4147)

JasonTheAdams commented on src/Providers/AiProviderRegistry.php:
```diff
@@ -0,0 +1,229 @@
+ *
+ * @since n.e.x.t
+ */
+class AiProviderRegistry
```
This should just be `ProviderRegistry`, without the `Ai` prefix.
```suggestion
class ProviderRegistry
```

JasonTheAdams replied:
I want to branch off of this, so I renamed this so I can proceed. Resolved in 06411b9

Ref34t commented:
@JasonTheAdams, all your comments are now resolved 
- All types and docblocks are fixed
- Moved the mock tests to the right directory
- Changed the `AiProviderRegistryTest ` to `ProviderRegistryTest`

You can have a look now. Thanks Alot 💯

felixarntz commented on src/Providers/ProviderRegistry.php:
```diff
@@ -0,0 +1,231 @@
+    public function hasProvider(string $idOrClassName): bool
+    {
+        return isset($this->providerClassNames[$idOrClassName]) ||
+            in_array($idOrClassName, $this->providerClassNames, true);
```
`in_array` is not as efficient as using `isset`. Since this method may be commonly called, I think it would be best to add another class property keyed by the provider class names. This way you can use `isset` for the second check here too.

Ref34t replied:
resolved here https://github.com/WordPress/php-ai-client/pull/38/commits/6e252404e38ea8f7f10f0ce44d9488b6e609f97b

felixarntz commented on src/Providers/ProviderRegistry.php:
```diff
@@ -0,0 +1,231 @@
+            throw new InvalidArgumentException(
+                sprintf('Provider class must implement %s: %s', ProviderInterface::class, $className)
+            );
+        }
```
This is unnecessary. It's impossible for `$className` to not implement `ProviderInterface`, since it's already enforced at the registration level, which is the only way to modify this property.

If PHPStan complains about this, we should rather ignore that because it's an invalid complaint I think.

Ref34t replied:
@felixarntz resolved here https://github.com/WordPress/php-ai-client/pull/38/commits/f868e7be407ed50b3ffc823d26fed89797b769eb

Ref34t commented:
@felixarntz thanks for those remarkable suggestions I appreciate it.  I'll update them today

Ref34t commented:
@felixarntz Both suggestions implemented as shown above. in your hands now 👍

JasonTheAdams commented:
Merged! Thanks again, @Ref34t! 🎉

Ref34t commented:
@JasonTheAdams awesome 😎 into the next one
